common/src/main/java/whocraft/tardis_refined/common/items/KeyItem.java

```diff
@@ -119,54 +141,58 @@ public InteractionResult interactLivingEntity(ItemStack itemStack, Player player
+            if (TardisRefined.KeySummonsItem) {
                 if (context.getPlayer().getAbilities().instabuild && context.getPlayer().isCrouching()) {
 
-                    var keychain = keychain(context.getItemInHand());
+                    var keychain = getKeychain(context.getItemInHand());
                     if (keychain.size() > 0) {
                         ResourceKey<Level> tardis = keychain.get(0);
                         var tardisLevel = Platform.getServer().levels.get(tardis);
                         TardisLevelOperator.get(tardisLevel).ifPresent(cap -> {
-                            if (cap.getControlManager().isInFlight()) {
-                                cap.getControlManager().setTargetPosition(context.getClickedPos().above());
-                                cap.getControlManager().getTargetLocation().rotation = context.getHorizontalDirection().getOpposite();
-                                cap.getControlManager().endFlight();
-                                PlayerUtil.sendMessage(context.getPlayer(), "TARDIS LANDING", true);
-                            } else {
-                                cap.getControlManager().beginFlight();
-                                PlayerUtil.sendMessage(context.getPlayer(), "TARDIS TAKING OFF", true);
-                            }
+                            cap.getControlManager().setTargetPosition(context.getClickedPos().above());
+                            cap.getControlManager().getTargetLocation().rotation = context.getHorizontalDirection().getOpposite();
+                            cap.getControlManager().beginFlight(true);
+                            PlayerUtil.sendMessage(context.getPlayer(), "TARDIS ON ITS WAY", true);
                         });
                     }
                 }
+            } else {
+                if (context.getPlayer().isCrouching()) {
+                    var keychain = getKeychain(context.getItemInHand());
+                    if (keychain.size() > 0) {
+                        Collections.rotate(keychain.subList(0, keychain.size()), -1);
+                        setKeychain(context.getItemInHand(), keychain);
+                        context.getPlayer().displayClientMessage(Component.translatable(ModMessages.MSG_KEY_CYCLED, keychain.get(0).location().getPath()), true);
+                        context.getLevel().playSound(null, context.getPlayer().blockPosition(), SoundEvents.UI_BUTTON_CLICK, SoundSource.BLOCKS, 1,2);
+                    }
+                }
             }
         }
 
+
         return super.useOn(context);
     }
 
     @Override
     public void appendHoverText(ItemStack itemStack, @Nullable Level level, List<Component> list, TooltipFlag tooltipFlag) {
         super.appendHoverText(itemStack, level, list, tooltipFlag);
 
-        ArrayList<ResourceKey<Level>> keychain = KeyItem.keychain(itemStack);
+        ArrayList<ResourceKey<Level>> keychain = KeyItem.getKeychain(itemStack);
```

Comment:
Shall we have a field for the levels and reference that rather than reconstruct the object every time? This method will be getting called every tick by the item tool tip renderer.